rama-tls-acme/src/client.rs

```diff
@@ -0,0 +1,664 @@
+    }
+
+    /// Create [`KeyAuthorization`] for the given challenge
+    pub fn create_key_authorization(
+        &self,
+        challenge: &server::Challenge,
+    ) -> Result<KeyAuthorization, OpaqueError> {
+        KeyAuthorization::new(&challenge.token, &self.account.credentials.key.create_jwk())
+    }
+
+    /// Create challenge data need for tls-alpn challenge
+    ///
+    /// This function returns a private private key and certificate which a TLS
+    /// backend should expose on port 443 (on the configured domain)
+    pub fn create_tls_challenge_data(
+        &self,
+        challenge: &server::Challenge,
+        identifier: &common::Identifier,
+    ) -> Result<(PrivatePkcs8KeyDer<'_>, Certificate), OpaqueError> {
+        let key_authz = self.create_key_authorization(challenge)?;
+
+        let mut cert_params = rcgen::CertificateParams::new(vec![identifier.clone().into()])
+            .context("create certificate params")?;
+        cert_params.extended_key_usages = vec![rcgen::ExtendedKeyUsagePurpose::ServerAuth];
+        cert_params.custom_extensions = vec![rcgen::CustomExtension::new_acme_identifier(
+            key_authz.digest().as_ref(),
+        )];
+
+        let key_pair = rcgen::KeyPair::generate().context("generate temporary keypair")?;
+        let key_der = key_pair.serialize_der();
+
+        let cert = cert_params
+            .self_signed(&key_pair)
+            .context("sign certificate params")?;
+        let pk = PrivatePkcs8KeyDer::from(key_der);
+        Ok((pk, cert))
+    }
+
+    async fn post(
+        &self,
+        url: &str,
+        payload: Option<&impl Serialize>,
+    ) -> Result<Response, OpaqueError> {
+        self.account.post(url, payload).await
+    }
+}
+
+async fn parse_response<T: serde::de::DeserializeOwned + Send + 'static>(
+    response: Response,
+) -> Result<T, ClientError> {
+    let body = response.into_body();
+    let bytes = body.collect().await.context("collect body")?.to_bytes();
+
+    let result = serde_json::from_slice::<T>(&bytes);
+    match result {
+        Ok(result) => Ok(result),
+        Err(err) => {
+            let problem = serde_json::from_slice::<server::Problem>(&bytes);
+            match problem {
+                Ok(problem) => Err(problem.into()),
+                Err(_err) => Err(err.context("parse problem response").into()),
+            }
+        }
+    }
+}
+
+fn get_retry_after_duration(response: &Response) -> Option<Duration> {
+    response
+        .headers()
+        .typed_get::<RetryAfter>()
+        .and_then(|after| match after.after() {
+            rama_http::headers::After::DateTime(http_date) => SystemTime::from(http_date)
+                .duration_since(SystemTime::now())
+                .ok(),
+            rama_http::headers::After::Delay(seconds) => Some(Duration::from(seconds)),
+        })
+}
+
+async fn retry_bad_nonce<F, Fut, T>(do_request: F) -> Result<T, ClientError>
+where
+    F: Fn() -> Fut,
```

Comment:
unless you have strong reason not to I usually use `FnOnce + Clone` as it still allows `Fn` but allows a bit more